File: Android/doric/src/main/java/com/github/penfeizhou/doric/DoricContext.java
package com.github.penfeizhou.doric;

import android.content.Context;
import android.view.ViewGroup;

import com.github.penfeizhou.doric.async.AsyncResult;
import com.github.penfeizhou.doric.plugin.DoricJavaPlugin;
import com.github.penfeizhou.doric.utils.DoricConstant;
import com.github.penfeizhou.doric.utils.DoricMetaInfo;
import com.github.penfeizhou.doric.shader.RootNode;
import com.github.pengfeizhou.jscore.JSDecoder;
import com.github.pengfeizhou.jscore.JSONBuilder;

import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

/**
 * @Description: Doric
 * @Author: pengfei.zhou
 * @CreateDate: 2019-07-18
 */
public class DoricContext {
    private final String mContextId;
    private final Map<String, DoricJavaPlugin> mPluginMap = new HashMap<>();
    private final Context mContext;
    private RootNode mRootNode = new RootNode(this);
    private final String source;
    private String script;
    private JSONObject initParams;

    DoricContext(Context context, String contextId, String source) {
        this.mContext = context;
        this.mContextId = contextId;
        this.source = source;
    }

    public String getSource() {
        return source;
    }

    public String getScript() {
        return script;
    }

    public static DoricContext create(Context context, String script, String source) {
        DoricContext doricContext = DoricContextManager.getInstance().createContext(context, script, source);
        doricContext.script = script;
        doricContext.callEntity(DoricConstant.DORIC_ENTITY_CREATE);
        return doricContext;
    }

    public void init(int width, int height) {
        this.initParams = new JSONBuilder()
                .put("width", width)
                .put("height", height).toJSONObject();
        callEntity(DoricConstant.DORIC_ENTITY_INIT, this.initParams);
    }

    public AsyncResult<JSDecoder> callEntity(String methodName, Object... args) {
        return getDriver().invokeContextEntityMethod(mContextId, methodName, args);
    }

    public IDoricDriver getDriver() {
        return DoricDriver.getInstance();
    }

    public RootNode getRootNode() {
        return mRootNode;
    }

    public Context getContext() {
        return mContext;
    }

    public String getContextId() {
        return mContextId;
    }

    public void teardown() {
        callEntity(DoricConstant.DORIC_ENTITY_DESTROY);
        DoricContextManager.getInstance().destroyContext(this).setCallback(new AsyncResult.Callback<Boolean>() {
            @Override
            public void onResult(Boolean result) {

            }

            @Override
            public void onError(Throwable t) {

            }

            @Override
            public void onFinish() {
                mPluginMap.clear();
            }
        });
    }

    public DoricJavaPlugin obtainPlugin(DoricMetaInfo<DoricJavaPlugin> doricMetaInfo) {
        DoricJavaPlugin plugin = mPluginMap.get(doricMetaInfo.getName());
        if (plugin == null) {
            plugin = doricMetaInfo.createInstance(this);
            mPluginMap.put(doricMetaInfo.getName(), plugin);
        }
        return plugin;
    }

    public void reload(String script) {
        getDriver().createContext(mContextId, script, source);
        callEntity(DoricConstant.DORIC_ENTITY_INIT, this.initParams);
    }

    public void onShow() {
        callEntity(DoricConstant.DORIC_ENTITY_SHOW);
    }

    public void onHidden() {
        callEntity(DoricConstant.DORIC_ENTITY_HIDDEN);
    }
}

File: doric-flutter/doric/lib/doric/doric_context.dart
import 'dart:async';
import 'dart:convert';
import 'dart:ffi';

import 'package:flutter/widgets.dart';
import 'package:doric/doric/doric_registry.dart';
import 'package:doric/doric/engine/doric_constant.dart';
import 'package:doric/doric/engine/doric_engine.dart';
import 'package:doric/doric/shader/doric_root.dart';
import 'package:doric/doric/shader/view_node.dart';
import 'package:doric/doric/utils/log.dart';
import 'package:doric/doric/utils/util.dart';
import 'package:doric/jscore/js_value.dart';

import '../flutter_jscore.dart';
import 'engine/doric_native_driver.dart';
import 'engine/doric_plugin.dart';

class DoricContextManager {
  int counter = 0;
  Map<String, DoricContext> contextMap = {};

  DoricContextManager._();

  static final _instance = DoricContextManager._();

  factory DoricContextManager.getInstance() => _instance;

  DoricContext createContext(
      final String script, final String source, String extra) {
    String contextId = counter.toString();
    DoricContext doricContext = DoricContext(contextId, source, extra);
    contextMap[contextId] = doricContext;
    counter++;
    doricContext.getDriver().createContext(contextId, script, source);
    return doricContext;
  }

  DoricContext getTopContext() {
    return contextMap.values.last;
  }

  bool destroyContext(DoricContext doricContext) {
    var success =
        doricContext.getDriver().destroyContext(doricContext.getContextId());
    contextMap.remove(doricContext.getContextId());
    return success;
  }

  void removeAll() {
    for (var i = 0; i < aliveContexts().length; i++) {
      aliveContexts().toList()[i].teardown();
    }
  }

  static DoricContext getContext(String contextId) {
    return DoricContextManager.getInstance().contextMap[contextId];
  }

  static Iterable<DoricContext> aliveContexts() {
    return DoricContextManager.getInstance().contextMap.values;
  }
}

class DoricContext {
  String mContextId;
  String source;
  String script;
  String extra;
  Object initParams;
  Map<String, DoricPlugin> mPluginMap = {};
  Map<String, Map<String, ViewNode>> mHeadNodes = {}; //
  RootNode root;
  bool isTeardown = false;

  DoricContext(this.mContextId, this.source, this.extra) {
    root = new RootNode(this);
  }

  AnimationController animationController;

  Iterable<ViewNode> allHeadNodes(String type) {
    return Map<String, ViewNode>.from((mHeadNodes[type] ?? {})).values;
  }

  ViewNode removeHeadNode(String type, String id) {
    return (mHeadNodes[type] ?? {}).remove(id);
  }

  void addHeadNode(String type, ViewNode viewNode) {
    Map typeNodes = mHeadNodes[type];
    if (typeNodes == null) {
      typeNodes = Map<String, ViewNode>();
    }
    typeNodes[viewNode.getId()] = viewNode;
    mHeadNodes[type] = typeNodes;
  }

  ViewNode targetViewNode(String id) {
    if (id == root.getId()) {
      return root;
    }
    var headVal = mHeadNodes.values.toList();
    for (var value in headVal) {
      if (value.containsKey(id)) {
        return value[id];
      }
    }
    return null;
  }

  setAnimationController(AnimationController animationController) {
    this.animationController = animationController;
  }

  AnimationController getAnimationController() {
    return this.animationController;
  }

  void init(String initData) {
    this.extra = initData;
    if (initData != null) {
      callEntity(DoricConstant.DORIC_ENTITY_INIT,
          [JSValue.makeString(getJSContext(), initData)]);
    }
  }

  static DoricContext create(String script, String source, String extra) {
    DoricContext doricContext =
        DoricContextManager.getInstance().createContext(script, source, extra);
    doricContext.script = script;
    doricContext.initContext();
    return doricContext;
  }

  initContext() {
    Timer.run(() {
      init(extra);
      callEntity(DoricConstant.DORIC_ENTITY_CREATE, []);
      build(DoricUtils.getScreenWidth(), DoricUtils.getScreenHeight());
      onShow();
    });
  }

  BuildContext getContext() {
    return root.getLayer().getContext();
  }

  void build(double width, double height) {
    print(width.toString()+"   "+height.toString());
    callEntity(DoricConstant.DORIC_ENTITY_BUILD, [
      JSValue.makeFromObject(
          getJSContext(),
          initParams = {
            "width": width,
            "height": height,
          })
    ]);
  }

  void onShow() {
    callEntity(DoricConstant.DORIC_ENTITY_SHOW, []);
  }

  void onHidden() {
    callEntity(DoricConstant.DORIC_ENTITY_HIDDEN, []);
  }

  void teardown() {
    DoricLog.i("teardown");
    isTeardown = true;
    callEntity(DoricConstant.DORIC_ENTITY_DESTROY, []);
    mPluginMap.forEach((key, value) {
      value.onTearDown();
    });
    mPluginMap.clear();
    DoricContextManager.getInstance().destroyContext(this);
  }

  DoricPlugin obtainPlugin(String name, DoricPluginCreator creator) {
    DoricPlugin plugin = mPluginMap[name];
    if (plugin != null) {
      return plugin;
    }
    plugin = creator.create(this);
    mPluginMap[name] = plugin;
    return plugin;
  }

  RootNode getRoot() {
    return root;
  }

  getContextId() {
    return mContextId;
  }

  JSValue callEntity(String methodName, List<JSValue> args) {
    return getDriver().invokeContextEntityMethod(mContextId, methodName, args);
  }

  JSContext getJSContext() {
    return getDriver().getJSContext();
  }

  DoricNativeDriver getDriver() {
    return DoricNativeDriver.getInstance();
  }
}

class DoricContextHolder {
  final DoricContext _context;

  DoricContextHolder(this._context);

  DoricContext getDoricContext() {
    return _context;
  }
}
